indexing-service/src/main/java/org/apache/druid/indexing/common/task/IndexTask.java

```diff
@@ -935,7 +900,7 @@ private TaskStatus generateAndPublishSegments(
         final BatchAppenderatorDriver driver = newDriver(appenderator, toolbox, segmentAllocator);
         final Firehose firehose = firehoseFactory.connect(dataSchema.getParser(), firehoseTempDir)
     ) {
-      this.appenderator = appenderator;
+      registerResourceCloserOnAbnormalExit(config -> appenderator.closeNow());
```

Comment:
hm, I just saw that Appenderator has the following javadoc:

```
Concurrency: all methods defined in this class directly, including {@link #close()} and {@link #closeNow()}, i. e.
 * all methods of the data appending and indexing lifecycle except {@link #drop} must be called from a single thread.
```

Maybe we should move the `appenderator.closeNow()` calls to an interrupt handler block in the main runTask methods and let interrupts through the resource closer trigger that.

Some of the appenderators are created in try-with, but AppenderatorImpl close() checks if it's already closed, so calling close() after closeNow() should be fine.

Reply:
Good point. Will update the pr soon.

Reply:
Took `Closeable` out of `Appenderator`. Callers of `Appenderator` should explicitly call `close()` or `closeNow()`.